python/ray/data/_internal/execution/operators/map_operator_tasks_impl.py

```diff
@@ -0,0 +1,184 @@
+        # Execution arguments.
+        self._op = op
+        self._transform_fn = op.get_transform_fn()
+        self._ray_remote_args = op.ray_remote_args()
+
+        # The temporary block bundle used to accumulate inputs until they meet the
+        # min_rows_per_batch requirement.
+        self._block_bundle: Optional[RefBundle] = None
+        self._min_rows_per_batch: int = op._min_rows_per_batch
+
+        # Execution state.
+        self._tasks: Dict[ObjectRef[ObjectRefGenerator], _TaskState] = {}
+        self._tasks_by_output_order: Dict[int, _TaskState] = {}
+        self._input_deps_done: bool = 0
+        self._next_task_index: int = 0
+        self._next_output_index: int = 0
+        self._obj_store_mem_alloc: int = 0
+        self._obj_store_mem_freed: int = 0
+        self._obj_store_mem_cur: int = 0
+        self._obj_store_mem_peak: int = 0
+
+    def add_input(self, bundle: RefBundle) -> None:
+        if self._min_rows_per_batch is None:
+            self._create_task(bundle)
+            return
+
+        def get_num_rows(bundle: Optional[RefBundle]):
+            if bundle is None:
+                return 0
+            if bundle.num_rows() is None:
+                return float("inf")
+            return bundle.num_rows()
+
+        bundle_rows = get_num_rows(bundle)
+        if bundle_rows == 0:
+            return
+
+        num_rows = get_num_rows(self._block_bundle) + bundle_rows
+        if num_rows > self._min_rows_per_batch:
+            if self._block_bundle:
+                self._create_task(self._block_bundle)
+                self._block_bundle = bundle
+            else:
+                self._create_task(bundle)
+        else:
+            self._block_bundle = merge_ref_bundles(self._block_bundle, bundle)
+
+    def inputs_done(self, input_index: int) -> None:
+        self._input_deps_done += 1
+        assert self._input_deps_done <= len(self._op._input_dependencies)
+        if (
+            self._input_deps_done == len(self._op._input_dependencies)
+            and self._block_bundle
+        ):
+            self._create_task(self._block_bundle)
+            self._block_bundle = None
+
+    def work_completed(self, ref: ObjectRef[ObjectRefGenerator]) -> None:
+        task = self._tasks.pop(ref)
+        all_refs = list(ray.get(ref))
+        block_refs = all_refs[:-1]
+        block_metas = ray.get(all_refs[-1])
+        del ref
+        assert len(block_metas) == len(block_refs), (block_refs, block_metas)
+        for ref in block_refs:
+            trace_allocation(ref, "map_operator_work_completed")
+        task.output = RefBundle(list(zip(block_refs, block_metas)), owns_blocks=True)
+        allocated = task.output.size_bytes()
+        self._obj_store_mem_alloc += allocated
+        self._obj_store_mem_cur += allocated
+        # TODO(ekl) this isn't strictly correct if multiple operators depend on this
+        # bundle, but it doesn't happen in linear dags for now.
```

Comment:
Do we have an idea of how to address this TODO?

Reply:
Yeah, we have to add a "manual" ref count to the block bundle.